### PR TITLE
Phase 5

### LINKED ISSUE #37: `ConnectionReader::host_consensus_state()` is not implemented
The `ConnectionReader::host_consensus_state()` impl is currently returning an empty mock result -> https://github.com/informalsystems/basecoin-rs/blob/main/src/app/modules/ibc.rs#L341-L350
```rust
fn host_consensus_state(
    &self,
    _height: IbcHeight,
) -> Result<AnyConsensusState, ConnectionError> {
    // FIXME: store host consensus state and return it here
    Ok(AnyConsensusState::Tendermint(ConsensusState::new(
        CommitmentRoot::from_bytes(&[]),
        Time::unix_epoch(),
        Hash::None,
    )))
}
```
This results in connection handhake verification failure (there are other problems with ibc-rs integration that I'm yet to figure out but this is one of them). 

The reason this call was left unimplemented is because Tendermint doesn't provide a way for the ABCI app to query the consensus engine. (see https://github.com/tendermint/tendermint/discussions/5977)
The way the Cosmos-SDK gets around this is by having the staking module save consensus states (to private store) while handling the ABCI `BeginBlock` method. (see https://github.com/cosmos/cosmos-sdk/blob/master/docs/architecture/adr-017-historical-header-module.md)

I propose a slightly different approach for basecoin - which is to extend the modules trait to add a `begin_block` method that'll allow the `ibc` module to store consensus states in its private store.

### PR DESCRIPTION
**Pull request by hu55a1n1**
Implements Phase 5 of #1 - Add support IBC Channel machinery
(Resolves #37)

- [x] complete/import the ics04 machinery from ibc-rs
- [x] state will now need to be able to store channel ends and packet data
- [x] should now be able to manually complete channel handshakes between two chains and send packets

### Testing
1. Follow the steps outlined in the README.md to set up a test environment running a basecoin chain (i.e. `basecoin-0`) and two ibc chains (`ibc-0` & `ibc-1`). 
2. Run the following command to create a connection between  `ibc-0` and `basecoin-0` -> 
```sh
$ RUST_BACKTRACE=1 hermes create channel ibc-0 basecoin-0 --port-a transfer --port-b transfer
# ...
Success: Channel {
    ordering: Unordered,
    a_side: ChannelSide {
        chain: ProdChainHandle {
            chain_id: ChainId {
                id: "ibc-0",
                version: 0,
            },
            runtime_sender: Sender { .. },
        },
        client_id: ClientId(
            "07-tendermint-1",
        ),
        connection_id: ConnectionId(
            "connection-1",
        ),
        port_id: PortId(
            "transfer",
        ),
        channel_id: Some(
            ChannelId(
                "channel-1",
            ),
        ),
    },
    b_side: ChannelSide {
        chain: ProdChainHandle {
            chain_id: ChainId {
                id: "basecoin-0",
                version: 0,
            },
            runtime_sender: Sender { .. },
        },
        client_id: ClientId(
            "07-tendermint-1",
        ),
        connection_id: ConnectionId(
            "connection-1",
        ),
        port_id: PortId(
            "transfer",
        ),
        channel_id: Some(
            ChannelId(
                "channel-1",
            ),
        ),
    },
    connection_delay: 0ns,
    version: Some(
        "ics20-1",
    ),
}
```
3. To create a channel in the opposite direction, run -> `RUST_BACKTRACE=1 hermes create channel basecoin-0 ibc-0 --port-a transfer --port-b transfer`

### CODE REVIEW

**Comment by hu55a1n1** [comment]
We're now depending on the basecoin/phase-5-1 branch of ibc-rs, which is basically the master branch with https://github.com/informalsystems/ibc-rs/pull/1711 & https://github.com/informalsystems/ibc-rs/pull/1746 merged. So we can depend on the next release once those two PRs are merged. 👍

File: src/app/modules/ibc.rs
Comment (thanethomson) [inline src/app/modules/ibc.rs:505]: For interest's sake, what does this entail? I tried poking around in https://github.com/cosmos/ibc-go but couldn't find the corresponding code (haven't spent a huge amount of time reading through it though, so I probably just missed it).

File: src/app/modules/ibc.rs
Comment (thanethomson) [inline src/app/modules/ibc.rs:518]: Just thinking about packet sequences in general in light of our previous conversation about Tendermint app concurrency, this for me makes me think that reading from/writing to a single channel/port will have to be synchronous, right? I mean, we could still write to multiple channel/port combinations in parallel, but a single channel/port will have to only be accessible from a single thread/coroutine?

Not something actionable right now, of course, just trying to think of the longer-term concurrency implications.

**Review by thanethomson — APPROVED** [review]
Overall looks good to me! Just a few questions inline.

I'm glad we have the integration test now - it was quite expensive previously to have to verify the implementation locally on my side :sweat_smile: I'm equally glad that the test passes, of course :grin:

File: src/app/mod.rs
Comment (hu55a1n1) [reply src/app/mod.rs:296]: Thanks for reviewing, @thanethomson! 🙏 
I replaced the `commit()` method with the `store()` method because some modules need to share their module stores with gRPC service instances. I am working on a better design for this on the `typed-stores` branch - I'll document this in a separate issue.

File: src/app/modules/ibc.rs
Comment (hu55a1n1) [reply src/app/modules/ibc.rs:505]: I was thinking we should do something like this -> https://github.com/cosmos/cosmos-sdk/blob/master/x/capability/keeper/keeper.go#L280 
Also, we might need to modify the API on the ibc-rs module side -> https://github.com/informalsystems/ibc-rs/issues/1769

File: src/app/modules/ibc.rs
Comment (hu55a1n1) [reply src/app/modules/ibc.rs:518]: This is an important observation and we should definitely keep this in mind. In the current design, modules have exclusive write access to their module stores. These stores are shared with the gRPC query services which only require read access. None of the message gRPC services have been implemented yet. So, although module implementations are required to be MT-safe, the module stores are never written in parallel because (AFAIK) tendermint ABCI `deliverTx` is never called in parallel.

**Pull request merged**
